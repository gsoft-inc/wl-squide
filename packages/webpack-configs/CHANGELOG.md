# @squide/webpack-configs

## 2.0.0

### Major Changes

- [#168](https://github.com/gsoft-inc/wl-squide/pull/168) [`89ace29`](https://github.com/gsoft-inc/wl-squide/commit/89ace29b9aeadbbe83cfa71dd137b9f1a115c283) Thanks [@patricklafrance](https://github.com/patricklafrance)! - This release Migrates Squide from Webpack Module Federation to [Module Federation 2.0](https://module-federation.io/guide/start/quick-start.html).

  This release deprecates the following packages:

  - `@squide/webpack-module-federation`, use `@squide/module-federation` instead.
  - `@squide/firefly-configs`, use `@squide/firefly-webpack-configs` instead.

  And introduce a few changes to existing API:

  - The `FireflyRuntime` nows accept a `useMsw` option and expose a new `isMswEnabled` getter:

  ```ts
  // bootstrap.tsx

  import { FireflyRuntime } from "@squide/firefly";

  const runtime = new FireflyRuntime({
    useMsw: true,
  });

  // Use the runtime to determine if MSW handlers should be registered.
  if (runtime.isMswEnabled) {
    // ...
  }
  ```

  - The `registerRemoteModules` function doesn't accept the remotes URL anymore. The remotes URL should be configured in the webpack configuration files.

  Previously:

  ```ts
  // bootstrap.tsx

  import {
    registerRemoteModules,
    type RemoteDefinition,
  } from "@squide/firefly";

  const Remotes: RemoteDefinition = [
    {
      name: "remote1",
      url: "http://localhost:8081",
    },
  ];

  await registerRemoteModules(Remotes, runtime);
  ```

  ```js
  // webpack.dev.js

  import { defineDevHostConfig } from "@squide/firefly-webpack-configs";
  import { swcConfig } from "./swc.dev.js";

  export default defineDevHostConfig(swcConfig, 8080, {
    overlay: false,
  });
  ```

  Now:

  ```ts
  // bootstrap.tsx

  import {
    registerRemoteModules,
    type RemoteDefinition,
  } from "@squide/firefly";

  const Remotes: RemoteDefinition = [
    {
      name: "remote1",
    },
  ];

  await registerRemoteModules(Remotes, runtime);
  ```

  ```js
  // webpack.dev.js

  import { defineDevHostConfig } from "@squide/firefly-webpack-configs";
  import { swcConfig } from "./swc.dev.js";

  /**
   * @typedef {import("@squide/firefly-webpack-configs").RemoteDefinition}[]
   */
  export const Remotes = [
    {
      name: "remote1",
      url: "http://localhost:8081",
    },
  ];

  export default defineDevHostConfig(swcConfig, 8080, Remotes, {
    overlay: false,
  });
  ```

  To migrate:

  1. Replace the `@squide/webpack-module-federation` dependency by `@squide/module-federation`.

  2. Replace the `@squide/firefly-configs` dependency by `@squide/firefly-webpack-configs`.

  3. Move the remotes URL from the `bootstrap.tsx` file to the `webpack.*.js` files.

  4. Integrate the new `useMsw` and `isMswEnabled` props.

## 1.2.8

### Patch Changes

- [#165](https://github.com/gsoft-inc/wl-squide/pull/165) [`c9358e5`](https://github.com/gsoft-inc/wl-squide/commit/c9358e5a276e38d877d16a6831251c1932b8497e) Thanks [@bsokol-wl](https://github.com/bsokol-wl)! - update webpack config dependency to remove hashes from chunk filenames

## 1.2.7

### Patch Changes

- [#162](https://github.com/gsoft-inc/wl-squide/pull/162) [`2c0840c`](https://github.com/gsoft-inc/wl-squide/commit/2c0840cb1c2c22e830e4f25f438c240c2d75d27a) Thanks [@patricklafrance](https://github.com/patricklafrance)! - Moved @workleap/webpack-configs as a dependency.

## 1.2.6

### Patch Changes

- [#160](https://github.com/gsoft-inc/wl-squide/pull/160) [`535b62c`](https://github.com/gsoft-inc/wl-squide/commit/535b62c9ce72bf32b69f018d9467a18186d123a8) Thanks [@bsokol-wl](https://github.com/bsokol-wl)! - Add webpack memory caching by default

## 1.2.5

### Patch Changes

- [#154](https://github.com/gsoft-inc/wl-squide/pull/154) [`e440515`](https://github.com/gsoft-inc/wl-squide/commit/e4405150a3c364fd4029c345399891614a434176) Thanks [@patricklafrance](https://github.com/patricklafrance)! - Updated packages.

## 1.2.4

### Patch Changes

- [#152](https://github.com/gsoft-inc/wl-squide/pull/152) [`d27fe71`](https://github.com/gsoft-inc/wl-squide/commit/d27fe717f899e395c3f01af86aac3e015159d719) Thanks [@patricklafrance](https://github.com/patricklafrance)! - Updated dependency versions.

## 1.2.3

### Patch Changes

- [#148](https://github.com/gsoft-inc/wl-squide/pull/148) [`a448347`](https://github.com/gsoft-inc/wl-squide/commit/a4483478bb8b7ef1f24513244e8c2410bdb86bc1) Thanks [@patricklafrance](https://github.com/patricklafrance)! - Removed `@squide/webpack-module-federation` dependency.

## 1.2.2

### Patch Changes

- Updated dependencies []:
  - @squide/webpack-module-federation@3.0.5

## 1.2.1

### Patch Changes

- [#142](https://github.com/gsoft-inc/wl-squide/pull/142) [`524be12b`](https://github.com/gsoft-inc/wl-squide/commit/524be12b26fdde5fe2b5d95ab20e1167a2b812f1) Thanks [@patricklafrance](https://github.com/patricklafrance)! - Removing `i18next-browser-languagedetector` from the shared dependencies as it causing an Error at runtime.

## 1.2.0

### Minor Changes

- [#140](https://github.com/gsoft-inc/wl-squide/pull/140) [`6eaf4ac3`](https://github.com/gsoft-inc/wl-squide/commit/6eaf4ac3f6ac88b62045ce280562a5887589026b) Thanks [@patricklafrance](https://github.com/patricklafrance)! - Added i18next libraries as shared singleton dependencies

## 1.1.4

### Patch Changes

- [`2f077119`](https://github.com/gsoft-inc/wl-squide/commit/2f0771194fd2034602955b8d3f72e1cc43e20e64) Thanks [@patricklafrance](https://github.com/patricklafrance)! - The `defineDevHostConfig` function now enables `fastRefresh`.

## 1.1.3

### Patch Changes

- Updated dependencies [[`1cda1be`](https://github.com/gsoft-inc/wl-squide/commit/1cda1be30779d1a1d5d2e21eac043baff20c0f7e)]:
  - @squide/webpack-module-federation@3.0.4

## 1.1.2

### Patch Changes

- Updated dependencies []:
  - @squide/webpack-module-federation@3.0.3

## 1.1.1

### Patch Changes

- Updated dependencies [[`4c3b6f1`](https://github.com/gsoft-inc/wl-squide/commit/4c3b6f1929364844dda6c1190fc45c3b037e8df9)]:
  - @squide/webpack-module-federation@3.0.2

## 1.1.0

### Minor Changes

- [#115](https://github.com/gsoft-inc/wl-squide/pull/115) [`568255a`](https://github.com/gsoft-inc/wl-squide/commit/568255a50a519e7d19c8c2b03909559686cd24c4) Thanks [@patricklafrance](https://github.com/patricklafrance)! - - A new `features` option is now available with the `defineConfig` functions to add the `@squide/i18next` package to the shared dependencies.

### Patch Changes

- Updated dependencies []:
  - @squide/webpack-module-federation@3.0.1

## 1.0.0

### Major Changes

- [#112](https://github.com/gsoft-inc/wl-squide/pull/112) [`a9dda1c`](https://github.com/gsoft-inc/wl-squide/commit/a9dda1c3b010f616556fc3313c1934e20a26bc11) Thanks [@patricklafrance](https://github.com/patricklafrance)! - This is a new package that includes the webpack configs. Moving the webpack configs to a standalone projects allow `@squide/firefly` to take an optionnal dependency on `webpack`. Thisis useful for local module projects that use the firefly stack but don't want to setup webpack because they do not need to support an isolated development environment.

  This new packages includes:

  - `defineDevHostConfig`
  - `defineBuildHostConfig`
  - `defineDevRemoteModuleConfig`
  - `defineBuildRemoteModuleConfig`

### Patch Changes

- Updated dependencies [[`a9dda1c`](https://github.com/gsoft-inc/wl-squide/commit/a9dda1c3b010f616556fc3313c1934e20a26bc11)]:
  - @squide/webpack-module-federation@3.0.0
