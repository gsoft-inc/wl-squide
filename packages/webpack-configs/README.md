# @squide/webpack-configs

## Usage

View the [user's documentation](https://gsoft-inc.github.io/wl-squide/).

## ü§ù Contributing

View the [contributor's documentation](../../CONTRIBUTING.md).

## License

Copyright ¬© 2023, Workleap. This code is licensed under the Apache License, Version 2.0. You may obtain a copy of this license at https://github.com/gsoft-inc/workleap-license/blob/master/LICENSE.
