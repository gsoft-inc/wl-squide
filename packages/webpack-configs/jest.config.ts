import type { Config } from "jest";
import path from "node:path";
import { swcConfig } from "./swc.jest.ts";

const config: Config = {
    testEnvironment: "jsdom",
    transformIgnorePatterns: [
        // Must exclude @workleap/webpack-configs from the transform ignore files
        // because it's an ESM only project which must be processed by SWC.
        // The pattern is optimized for PNPM, for more info view:
        // - https://jestjs.io/docs/configuration#transformignorepatterns-arraystring
        // - https://jestjs.io/docs/ecmascript-modules
        `${path.join(
            __dirname,
            "../.."
        )}/node_modules/.pnpm/(?!(@workleap\\+webpack-configs)@)`
    ],
    transform: {
        // Must add the ".js" file extension because the files from @workleap/webpack-configs
        // are ESM only and therefore must be processed by SWC.
        "^.+\\.(js|ts)$": ["@swc/jest", swcConfig as Record<string, unknown>]
    },
    cacheDirectory: "./node_modules/.cache/jest"
};

export default config;
