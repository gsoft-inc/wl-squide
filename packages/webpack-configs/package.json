{
    "name": "@squide/webpack-configs",
    "author": "Workleap",
    "version": "1.2.5",
    "description": "Utilities to configure webpack with a @squide federated application shell.",
    "license": "Apache-2.0",
    "repository": {
        "type": "git",
        "url": "git+https://github.com/gsoft-inc/wl-squide.git",
        "directory": "packages/webpack-configs"
    },
    "publishConfig": {
        "access": "public",
        "provenance": true
    },
    "type": "module",
    "exports": {
        ".": {
            "import": "./dist/index.js",
            "types": "./dist/index.d.ts",
            "default": "./dist/index.js"
        }
    },
    "files": [
        "/dist",
        "CHANGELOG.md",
        "README.md"
    ],
    "scripts": {
        "dev": "tsup --config ./tsup.dev.ts",
        "build": "tsup --config ./tsup.build.ts"
    },
    "peerDependencies": {
        "@workleap/swc-configs": "*",
        "@workleap/webpack-configs": "*",
        "webpack": ">=5.0.0"
    },
    "devDependencies": {
        "@swc/core": "1.4.7",
        "@swc/jest": "0.2.36",
        "@testing-library/react": "14.2.1",
        "@types/jest": "29.5.12",
        "@types/node": "20.11.27",
        "@workleap/eslint-plugin": "3.1.0",
        "@workleap/swc-configs": "2.2.3",
        "@workleap/tsup-configs": "3.0.4",
        "@workleap/typescript-configs": "3.0.2",
        "@workleap/webpack-configs": "1.4.1",
        "jest": "29.7.0",
        "jest-environment-jsdom": "29.7.0",
        "ts-jest": "29.1.2",
        "tsup": "8.0.2",
        "typescript": "5.4.2",
        "webpack": "5.90.3"
    },
    "dependencies": {
        "deepmerge": "4.3.1",
        "html-webpack-plugin": "5.6.0"
    },
    "sideEffects": false,
    "engines": {
        "node": ">=20.0.0"
    }
}
