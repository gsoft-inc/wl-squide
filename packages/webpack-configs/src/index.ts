export * from "@workleap/webpack-configs";

export * from "./defineConfig.ts";
