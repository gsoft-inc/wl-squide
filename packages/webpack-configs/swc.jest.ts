import { defineJestConfig } from "@workleap/swc-configs";

// const commonJsTransformer: SwcConfigTransformer = config => {
//     return {
//         ...config,
//         module: {
//             // The output module resolution system that the code will be compiled for.
//             type: "commonjs"
//         }
//     };
// };

export const swcConfig = defineJestConfig();
