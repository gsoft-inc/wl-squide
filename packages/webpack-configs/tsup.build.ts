import { defineBuildConfig } from "@workleap/tsup-configs";

export default defineBuildConfig({
    // Temporary fix to support using "node:path".
    platform: undefined,
    entry: [
        "./src/index.ts",
        "./src/sharedDependenciesResolutionPlugin.ts",
        "./src/nonCacheableRemoteEntryPlugin.ts"
    ]
});
