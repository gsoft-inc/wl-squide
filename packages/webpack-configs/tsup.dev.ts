import { defineDevConfig } from "@workleap/tsup-configs";

export default defineDevConfig({
    // Temporary fix to support using "node:path".
    platform: undefined,
    // treeshake: true,
    entry: [
        "./src/index.ts",
        "./src/sharedDependenciesResolutionPlugin.ts",
        "./src/nonCacheableRemoteEntryPlugin.ts"
    ]
});
