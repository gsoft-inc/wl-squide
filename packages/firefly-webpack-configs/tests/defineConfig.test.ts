import { ModuleFederationPlugin } from "@module-federation/enhanced/webpack";
import type { ModuleFederationPluginOptions } from "@squide/webpack-configs";
import { defineBuildConfig as defineSwcBuildConfig, defineDevConfig as defineSwcDevConfig } from "@workleap/swc-configs";
import { findPlugin, matchConstructorName } from "@workleap/webpack-configs";
import type { WebpackPluginInstance } from "webpack";
import { defineBuildHostConfig, defineBuildRemoteModuleConfig, defineDevHostConfig, defineDevRemoteModuleConfig } from "../src/index.ts";

// The following options are relative to the environment running the test and breaks on CI.
function prepareModuleFederationPluginForSnapshot(plugin: WebpackPluginInstance) {
    delete (plugin._options as ModuleFederationPluginOptions)["runtimePlugins"];

    return plugin;
}

describe("defineDevHostConfig", () => {
    const SwcConfig = defineSwcDevConfig({
        chrome: "116"
    });

    test("includes react-router and msw dependencies", () => {
        const config = defineDevHostConfig(SwcConfig, 8080, []);
        const result = findPlugin(config, matchConstructorName(ModuleFederationPlugin.name));

        expect(prepareModuleFederationPluginForSnapshot(result.plugin as WebpackPluginInstance)).toMatchSnapshot();
    });
});

describe("defineBuildHostConfig", () => {
    const SwcConfig = defineSwcBuildConfig({
        chrome: "116"
    });

    test("includes react-router and msw dependencies", () => {
        const config = defineBuildHostConfig(SwcConfig, []);
        const result = findPlugin(config, matchConstructorName(ModuleFederationPlugin.name));

        expect(prepareModuleFederationPluginForSnapshot(result.plugin as WebpackPluginInstance)).toMatchSnapshot();
    });
});

describe("defineDevRemoteModuleConfig", () => {
    const SwcConfig = defineSwcDevConfig({
        chrome: "116"
    });

    test("includes react-router and msw dependencies", () => {
        const config = defineDevRemoteModuleConfig(SwcConfig, "remote1", 8081);
        const result = findPlugin(config, matchConstructorName(ModuleFederationPlugin.name));

        expect(prepareModuleFederationPluginForSnapshot(result.plugin as WebpackPluginInstance)).toMatchSnapshot();
    });
});

describe("defineBuildRemoteModuleConfig", () => {
    const SwcConfig = defineSwcBuildConfig({
        chrome: "116"
    });

    test("includes react-router and msw dependencies", () => {
        const config = defineBuildRemoteModuleConfig(SwcConfig, "remote1");
        const result = findPlugin(config, matchConstructorName(ModuleFederationPlugin.name));

        expect(prepareModuleFederationPluginForSnapshot(result.plugin as WebpackPluginInstance)).toMatchSnapshot();
    });
});


