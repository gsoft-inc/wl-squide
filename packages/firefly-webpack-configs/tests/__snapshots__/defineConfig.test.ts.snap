// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineBuildHostConfig includes react-router and msw dependencies 1`] = `
ModuleFederationPlugin {
  "_options": {
    "dts": false,
    "manifest": false,
    "name": "host",
    "remotes": {},
    "shared": {
      "@squide/core": {
        "eager": true,
        "singleton": true,
      },
      "@squide/module-federation": {
        "eager": true,
        "singleton": true,
      },
      "@squide/msw": {
        "eager": true,
        "singleton": true,
      },
      "@squide/react-router": {
        "eager": true,
        "singleton": true,
      },
      "react": {
        "eager": true,
        "requiredVersion": undefined,
        "singleton": true,
      },
      "react-dom": {
        "eager": true,
        "singleton": true,
      },
      "react-router-dom": {
        "eager": true,
        "singleton": true,
      },
    },
  },
  "name": "ModuleFederationPlugin",
}
`;

exports[`defineBuildRemoteModuleConfig includes react-router and msw dependencies 1`] = `
ModuleFederationPlugin {
  "_options": {
    "dts": false,
    "exposes": {
      "./register": "./src/register",
    },
    "filename": "remoteEntry.js",
    "manifest": false,
    "name": "remote1",
    "shared": {
      "@squide/core": {
        "eager": undefined,
        "singleton": true,
      },
      "@squide/module-federation": {
        "eager": undefined,
        "singleton": true,
      },
      "@squide/msw": {
        "eager": undefined,
        "singleton": true,
      },
      "@squide/react-router": {
        "eager": undefined,
        "singleton": true,
      },
      "react": {
        "eager": undefined,
        "requiredVersion": undefined,
        "singleton": true,
      },
      "react-dom": {
        "eager": undefined,
        "singleton": true,
      },
      "react-router-dom": {
        "eager": undefined,
        "singleton": true,
      },
    },
  },
  "name": "ModuleFederationPlugin",
}
`;

exports[`defineDevHostConfig includes react-router and msw dependencies 1`] = `
ModuleFederationPlugin {
  "_options": {
    "dts": false,
    "manifest": false,
    "name": "host",
    "remotes": {},
    "shared": {
      "@squide/core": {
        "eager": true,
        "singleton": true,
      },
      "@squide/module-federation": {
        "eager": true,
        "singleton": true,
      },
      "@squide/msw": {
        "eager": true,
        "singleton": true,
      },
      "@squide/react-router": {
        "eager": true,
        "singleton": true,
      },
      "react": {
        "eager": true,
        "requiredVersion": undefined,
        "singleton": true,
      },
      "react-dom": {
        "eager": true,
        "singleton": true,
      },
      "react-router-dom": {
        "eager": true,
        "singleton": true,
      },
    },
  },
  "name": "ModuleFederationPlugin",
}
`;

exports[`defineDevRemoteModuleConfig includes react-router and msw dependencies 1`] = `
ModuleFederationPlugin {
  "_options": {
    "dts": false,
    "exposes": {
      "./register": "./src/register",
    },
    "filename": "remoteEntry.js",
    "manifest": false,
    "name": "remote1",
    "shared": {
      "@squide/core": {
        "eager": undefined,
        "singleton": true,
      },
      "@squide/module-federation": {
        "eager": undefined,
        "singleton": true,
      },
      "@squide/msw": {
        "eager": undefined,
        "singleton": true,
      },
      "@squide/react-router": {
        "eager": undefined,
        "singleton": true,
      },
      "react": {
        "eager": undefined,
        "requiredVersion": undefined,
        "singleton": true,
      },
      "react-dom": {
        "eager": undefined,
        "singleton": true,
      },
      "react-router-dom": {
        "eager": undefined,
        "singleton": true,
      },
    },
  },
  "name": "ModuleFederationPlugin",
}
`;
